consensus/src/shared_table/mod.rs

```diff
@@ -112,19 +117,24 @@ impl SharedTableInner {
 
 		// TODO: consider a strategy based on the number of candidate votes as well.
 		// only check validity if this wasn't locally proposed.
-		let extra_work = para_member
-			&& self.proposed_digest.as_ref().map_or(true, |d| d != digest)
-			&& self.checked_validity.insert(digest.clone());
+		let do_validation = para_member && match self.validated.entry(digest.clone()) {
+			Entry::Occupied(_) => false,
+			Entry::Vacant(entry) => {
+				entry.insert(None);
+				true
+			}
+		};
 
-		let work = if extra_work {
+		let work = if do_validation {
+			let fetch_incoming = router.fetch_incoming(summary.group_id);
 			match self.table.get_candidate(&digest) {
 				None => None, // TODO: handle table inconsistency somehow?
```

Comment:
i'll tackle this now -- it should either be a warning or a panic because it's something that's really not supposed to happen with the interface of the table. The error variant makes sense to me.